napari_clusters_plotter/_kmeans_clustering.py

```diff
@@ -1,158 +1,155 @@
+                self.kmeans_nr_clusters.value,
+                self.kmeans_nr_iterations.value
             )
 
-        button.clicked.connect(run_clicked)
-        run_widget.layout().addWidget(button)
+        run_button.clicked.connect(run_clicked)
+        update_button.clicked.connect(self.update_properties_list)
+        defaults_button.clicked.connect(partial(restore_defaults, self, DEFAULTS))
 
-        # adding all widgets to the layout
-        self.layout().addWidget(label_container)
-        self.layout().addWidget(choose_img_container)
-        self.layout().addWidget(choose_properties_container)
-        self.layout().addWidget(self.clust_method_container)
-        self.layout().addWidget(self.kmeans_settings_container)
-        self.layout().addWidget(self.kmeans_settings_container2)
-        self.layout().addWidget(run_widget)
-        self.layout().setSpacing(0)
+        # update measurements list when a new labels layer is selected
+        self.labels_select.changed.connect(self.update_properties_list)
 
         # go through all widgets and change spacing
         for i in range(self.layout().count()):
             item = self.layout().itemAt(i).widget()
             item.layout().setSpacing(0)
             item.layout().setContentsMargins(3, 3, 3, 3)
 
-        # hide widget for the selection of parameters for kmeans unless kmeans clustering method is chosen
-        self.clust_method_choice_list.currentIndexChanged.connect(self.change_kmeans_clustering)
+        # hide widget for the selection of parameters unless specific method is chosen
+        self.clust_method_choice_list.changed.connect(self.change_kmeans_clustering)
 
     def change_kmeans_clustering(self):
-        widgets_inactive(self.kmeans_settings_container, self.kmeans_settings_container2,
-                         active=self.clust_method_choice_list.currentText() == 'KMeans')
-
-    def get_selected_label(self):
-        index = self.label_list.currentIndex()
-        if index >= 0:
-            return self._available_labels[index]
-        return None
-
-    def update_label_list(self):
-        selected_layer = self.get_selected_label()
-        selected_index = -1
-
-        self._available_labels = []
-        self.label_list.clear()
-        i = 0
-        for layer in self.viewer.layers:
-            if isinstance(layer, napari.layers.Labels):
-                self._available_labels.append(layer)
-                if layer == selected_layer:
-                    selected_index = i
-                self.label_list.addItem(layer.name)
-                i = i + 1
-        self.label_list.setCurrentIndex(selected_index)
+        widgets_inactive(self.kmeans_settings_container_nr, self.kmeans_settings_container_iter,
+                         active=self.clust_method_choice_list.current_choice == 'KMeans')
 
     def update_properties_list(self):
-        selected_layer = self.get_selected_label()
+        selected_layer = self.labels_select.value
```

Comment:
so cool 😎

Reply:
all thanks to the smart person who gave me tips in issues section :)